src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp

```diff
@@ -57,5 +72,86 @@ template <size_t VolumeDim>
 double oscillation_indicator(const DataVector& data,
                              const Mesh<VolumeDim>& mesh) noexcept;
 
+// Compute the unnormalized nonlinear WENO weights. This is a fairly standard
+// choice of weights; see e.g., Eq. 3.9 of Zhong2013 or Eq. 3.6 of Zhu2016.
+inline double unnormalized_nonlinear_weight(
+    const double linear_weight, const double oscillation_indicator) noexcept {
+  return linear_weight / square(1.e-6 + oscillation_indicator);
+}
+
+// Compute the WENO weighted reconstruction of several polynomials, see e.g.,
+// Eq. 4.3 of Zhu2016. This is fairly standard, though different references can
+// differ in their choice of oscillation/smoothness indicator.
+template <typename Tag, size_t VolumeDim, typename TagsList>
+void reconstruct_from_weighted_sum(
+    const gsl::not_null<db::item_type<Tag>*> local_tensor,
+    const Mesh<VolumeDim>& mesh, const double neighbor_linear_weight,
+    const std::unordered_map<
+        std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>,
+        Variables<TagsList>,
+        boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>&
+        neighbor_vars) noexcept {
+  // Pre-allocate storage for the weights
```

Comment:
I don't understand, where's the memory being allocated?

Reply:
This was a bad choice of words on my part to describe premature, un-measured "optimization" :-/. I've moved the unordered_map back to a more restricted scope inside the loops.